deployment/terraform/batch.tf

```diff
@@ -0,0 +1,75 @@
+#
+# Batch resources
+#
+resource "aws_launch_template" "batch_gpu_container_instance" {
+  block_device_mappings {
+    device_name = "/dev/xvda"
+
+    ebs {
+      volume_size = 64
+      volume_type = "gp2"
+    }
+  }
+}
+
+# Pull the AMI ID for the latest Amazon ECS GPU-optimized AMI
+# https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-optimized_AMI.html#w520aac22c15c31b5
+data "aws_ssm_parameter" batch_gpu_ami_id {
+  name = "/aws/service/ecs/optimized-ami/amazon-linux-2/gpu/recommended/image_id"
+}
+
+resource "aws_batch_compute_environment" "gpu" {
+  depends_on = [aws_iam_role_policy_attachment.batch_policy]
+
+  compute_environment_name = "batch${var.project}GPUComputeEnvironment"
+  type                     = "MANAGED"
+  state                    = "ENABLED"
+  service_role             = aws_iam_role.container_instance_batch.arn
+
+  compute_resources {
```

Comment:
The existing manual CE has support for a `SPOT_CAPACITY_OPTIMIZED` allocation strategy. It looks like support for this feature in Terraform is in-progress, but not yet complete (https://github.com/terraform-providers/terraform-provider-aws/issues/10980).

We should ensure that not supporting this feature doesn't degrade the existing AWS Batch experience for the application. In addition, I think we should create an issue to add support for it once the feature is released via the AWS provider.

Reply:
I looked into utilizing the [`instance_market_options`](https://www.terraform.io/docs/providers/aws/r/launch_template.html#instance_market_options) parameter of `aws_launch_template` to see if we could replicate this allocation strategy outside of the CE.

It doesn't seem possible. For now, I think we'll have to stay with the default (lowest-price) allocation strategy.

An AWS [blog post](https://aws.amazon.com/blogs/compute/introducing-the-capacity-optimized-allocation-strategy-for-amazon-ec2-spot-instances/) explains the differences:

>As a result, the lowest-price allocation strategy is a good choice for workloads with a low cost of interruption that want the lowest possible prices, such as:
>- Time-insensitive workloads
>- Extremely transient workloads
>- Workloads that are easily check-pointed and restarted
>
> . . .
>
>The capacity-optimized allocation strategy may be especially useful for workloads with a high cost of interruption, including:
>- Big data and analytics
>- Image and media rendering
>- Machine learning
>- High performance computing

It seems like not supporting this feature **will** degrade the existing AWS Batch experience. However, we're using the lowest-price allocation strategy everywhere else and haven't seem to run into issues with one-off tasks.

The PR connected to the AWS provider issue was reviewed today so it seems there will be support for this feature soon. I see a couple paths forward:

- Put this PR on hold until terraform-providers/terraform-provider-aws#10894 is merged
- Create an issue to add support for it and accept a temporary trade off for infrastructure as code

What do you think?

Reply:
I'm fine with the latter for this one (looks like the PR was merged and is slated for the upcoming 2.44.0 release).